File: benchmark/zig/bench_parse_mvt.zig
const std = @import("std");
const vtzero = @import("vtzero");

const GeometryChecksumHandler = struct {
    checksum: u64 = 0,

    pub fn points_begin(self: *GeometryChecksumHandler, n: u32) void {
        self.checksum +%= n;
    }
    pub fn points_point(self: *GeometryChecksumHandler, p: vtzero.Point) void {
        self.checksum +%= @as(u64, @bitCast(@as(i64, p.x)));
        self.checksum +%= @as(u64, @bitCast(@as(i64, p.y)));
    }
    pub fn points_end(self: *GeometryChecksumHandler) void {
        self.checksum +%= 1;
    }
    pub fn linestring_begin(self: *GeometryChecksumHandler, n: u32) void {
        self.checksum +%= n;
    }
    pub fn linestring_point(self: *GeometryChecksumHandler, p: vtzero.Point) void {
        self.checksum +%= @as(u64, @bitCast(@as(i64, p.x)));
        self.checksum +%= @as(u64, @bitCast(@as(i64, p.y)));
    }
    pub fn linestring_end(self: *GeometryChecksumHandler) void {
        self.checksum +%= 3;
    }
    pub fn ring_begin(self: *GeometryChecksumHandler, n: u32) void {
        self.checksum +%= n;
    }
    pub fn ring_point(self: *GeometryChecksumHandler, p: vtzero.Point) void {
        self.checksum +%= @as(u64, @bitCast(@as(i64, p.x)));
        self.checksum +%= @as(u64, @bitCast(@as(i64, p.y)));
    }
    pub fn ring_end(self: *GeometryChecksumHandler, ring_type: vtzero.RingType) void {
        self.checksum +%= @intFromEnum(ring_type);
    }
    pub fn result(self: *GeometryChecksumHandler) u64 {
        return self.checksum;
    }
};

pub fn parseTileBytes(data: []const u8) !u64 {
    var checksum: u64 = 0;
    var tile = vtzero.VectorTile.init(data);
    checksum +%= @intCast(try tile.countLayers());
    while (try tile.nextLayer()) |layer| {
        var mut_layer = layer;
        checksum +%= mut_layer.name().len;
        while (try mut_layer.nextFeature()) |feature| {
            checksum +%= feature.id();
            checksum +%= @intCast(feature.numProperties());
            checksum +%= @intFromEnum(feature.geometryType());
        }
    }
    return checksum;
}

pub fn parseTileBytesDecode(data: []const u8) !u64 {
    var checksum: u64 = 0;
    var tile = vtzero.VectorTile.init(data);
    checksum +%= @intCast(try tile.countLayers());
    while (try tile.nextLayer()) |layer| {
        var mut_layer = layer;
        checksum +%= mut_layer.name().len;
        while (try mut_layer.nextFeature()) |feature| {
            checksum +%= feature.id();
            checksum +%= @intCast(feature.numProperties());
            checksum +%= @intFromEnum(feature.geometryType());
            var handler: GeometryChecksumHandler = .{};
            checksum +%= try vtzero.decodeGeometry(feature.geometry(), &handler);
        }
    }
    return checksum;
}

pub fn main(init: std.process.Init.Minimal) !void {
    const io = std.Io.Threaded.global_single_threaded.io();
    const alloc = std.heap.page_allocator;

    var args = try std.process.Args.Iterator.initAllocator(init.args, alloc);
    defer args.deinit();
    _ = args.next();

    var decode_mode = false;
    var iters: usize = 1;
    var paths: std.ArrayList([]const u8) = .empty;
    defer paths.deinit(alloc);

    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "--decode")) {
            decode_mode = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--iters")) {
            const n = args.next() orelse return error.MissingItersValue;
            iters = try std.fmt.parseInt(usize, n, 10);
            continue;
        }
        try paths.append(alloc, arg);
    }

    if (paths.items.len == 0) return error.MissingTilePath;

    var buffers = try alloc.alloc([]u8, paths.items.len);
    defer alloc.free(buffers);
    const cwd = std.Io.Dir.cwd();
    for (paths.items, 0..) |p, i| {
        buffers[i] = try cwd.readFileAlloc(io, p, alloc, .limited(std.math.maxInt(usize)));
    }
    defer {
        for (buffers) |b| alloc.free(b);
    }

    var checksum: u64 = 0;
    const t0 = std.Io.Clock.awake.now(io).nanoseconds;
    for (0..iters) |_| {
        for (buffers) |data| {
            checksum +%= if (decode_mode) try parseTileBytesDecode(data) else try parseTileBytes(data);
        }
    }
    const t1 = std.Io.Clock.awake.now(io).nanoseconds;
    const elapsed_ns: u64 = @intCast(t1 - t0);

    const mode = if (decode_mode) "parse+decode" else "parse-only";
    var stdout_w = std.Io.File.stdout().writerStreaming(io, &.{});
    try stdout_w.interface.print("{s}\t{x}\t{d}\n", .{ mode, checksum, elapsed_ns });
}

File: benchmark/cpp/bench_parse_mvt.cpp
#include <vtzero/vector_tile.hpp>
#include <vtzero/geometry.hpp>

#include <chrono>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

struct geometry_checksum_handler {
    std::uint64_t checksum = 0;

    void points_begin(std::uint32_t n) { checksum += n; }
    void points_point(vtzero::point p) {
        checksum += static_cast<std::uint64_t>(static_cast<std::int64_t>(p.x));
        checksum += static_cast<std::uint64_t>(static_cast<std::int64_t>(p.y));
    }
    void points_end() { checksum += 1; }
    void linestring_begin(std::uint32_t n) { checksum += n; }
    void linestring_point(vtzero::point p) {
        checksum += static_cast<std::uint64_t>(static_cast<std::int64_t>(p.x));
        checksum += static_cast<std::uint64_t>(static_cast<std::int64_t>(p.y));
    }
    void linestring_end() { checksum += 3; }
    void ring_begin(std::uint32_t n) { checksum += n; }
    void ring_point(vtzero::point p) {
        checksum += static_cast<std::uint64_t>(static_cast<std::int64_t>(p.x));
        checksum += static_cast<std::uint64_t>(static_cast<std::int64_t>(p.y));
    }
    void ring_end(vtzero::ring_type rt) {
        checksum += static_cast<std::uint64_t>(rt == vtzero::ring_type::outer ? 0 : rt == vtzero::ring_type::inner ? 1 : 2);
    }
    std::uint64_t result() const { return checksum; }
};

static std::uint64_t parse_tile_bytes(const std::string& data) {
    std::uint64_t checksum = 0;
    vtzero::vector_tile tile{data};
    checksum += static_cast<std::uint64_t>(tile.count_layers());
    while (auto layer = tile.next_layer()) {
        checksum += static_cast<std::uint64_t>(layer.name().size());
        while (auto feature = layer.next_feature()) {
            checksum += feature.id();
            checksum += static_cast<std::uint64_t>(feature.num_properties());
            checksum += static_cast<std::uint64_t>(feature.geometry_type());
        }
    }
    return checksum;
}

static std::uint64_t parse_tile_bytes_decode(const std::string& data) {
    std::uint64_t checksum = 0;
    vtzero::vector_tile tile{data};
    checksum += static_cast<std::uint64_t>(tile.count_layers());
    while (auto layer = tile.next_layer()) {
        checksum += static_cast<std::uint64_t>(layer.name().size());
        while (auto feature = layer.next_feature()) {
            checksum += feature.id();
            checksum += static_cast<std::uint64_t>(feature.num_properties());
            checksum += static_cast<std::uint64_t>(feature.geometry_type());
            geometry_checksum_handler handler;
            checksum += vtzero::decode_geometry(feature.geometry(), handler);
        }
    }
    return checksum;
}

int main(int argc, char** argv) {
    bool decode = false;
    std::size_t iters = 1;
    std::vector<std::string> paths;

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--decode") {
            decode = true;
            continue;
        }
        if (arg == "--iters") {
            if (i + 1 >= argc) {
                std::cerr << "missing value for --iters\n";
                return 1;
            }
            iters = static_cast<std::size_t>(std::stoull(argv[++i]));
            continue;
        }
        paths.push_back(arg);
    }

    if (paths.empty()) {
        std::cerr << "usage: bench-parse-mvt-cpp [--decode] [--iters N] <tile...>\n";
        return 1;
    }

    try {
        std::vector<std::string> datas;
        datas.reserve(paths.size());
        for (const auto& p : paths) {
            std::ifstream in(p, std::ios::binary);
            if (!in) throw std::runtime_error("failed to open: " + p);
            datas.emplace_back(std::istreambuf_iterator<char>(in), std::istreambuf_iterator<char>());
        }

        std::uint64_t checksum = 0;
        const auto t0 = std::chrono::steady_clock::now();
        for (std::size_t i = 0; i < iters; ++i) {
            for (const auto& data : datas) {
                checksum += decode ? parse_tile_bytes_decode(data) : parse_tile_bytes(data);
            }
        }
        const auto t1 = std::chrono::steady_clock::now();
        const auto elapsed_ns =
            std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

        std::cout << (decode ? "parse+decode" : "parse-only")
                  << '\t' << std::hex << checksum << std::dec
                  << '\t' << elapsed_ns << '\n';
    } catch (const std::exception& e) {
        std::cerr << e.what() << '\n';
        return 1;
    }
    return 0;
}
